extensions/positron-code-cells/src/documentManager.ts

```diff
@@ -0,0 +1,269 @@
+		}
+	}
+
+	public runCurrentAdvance(line?: number): void {
+		this.runCurrentCell(line);
+		this.goToNextCell(line);
+	}
+
+	public runPreviousCell(line?: number): void {
+		const cell = this.getPreviousCell(line);
+		if (cell) {
+			this.runCell(cell);
+			this.goToPreviousCell(line);
+		}
+	}
+
+	public runNextCell(line?: number): void {
+		const cell = this.getNextCell(line);
+		if (cell) {
+			this.runCell(cell);
+			this.goToNextCell(line);
+		}
+	}
+
+	public runCellsAbove(line?: number): void {
+		const end = this.getCurrentCellIndex(line);
+		for (let i = 0; i < end; i++) {
+			this.runCell(this.cells[i]);
+		}
+	}
+
+	public runCurrentAndBelow(line?: number): void {
+		const start = this.getCurrentCellIndex(line);
+		for (let i = start; i < this.cells.length; i++) {
+			this.runCell(this.cells[i]);
+		}
+	}
+
+	public runCellsBelow(line?: number): void {
+		const start = this.getCurrentCellIndex(line);
+		for (let i = start + 1; i < this.cells.length; i++) {
+			this.runCell(this.cells[i]);
+		}
+	}
+
+	public goToPreviousCell(line?: number): void {
+		const cell = this.getPreviousCell(line);
+		if (cell) {
+			this.goToCell(cell);
+		}
+	}
+
+	public goToNextCell(line?: number): void {
+		const cell = this.getNextCell(line);
+		if (cell) {
+			this.goToCell(cell);
+		}
+	}
+
+	public async insertCodeCell(line?: number): Promise<void> {
+		const location = this.getCurrentCell(line)?.range.end ?? this.editor.selection.active;
+		await this.editor.edit(editBuilder => {
+			editBuilder.insert(location, getParser(this.document.languageId)?.newCell() ?? '');
+		});
+		this.goToNextCell(location.line);
+	}
+}
+
+export function canHaveCells(document: vscode.TextDocument) {
+	return !IGNORED_SCHEMES.includes(document.uri.scheme) && supportedLanguageIds.includes(document.languageId);
+}
+
+export function reparseDocument(editor: vscode.TextEditor | undefined) {
+	if (editor && canHaveCells(editor.document)) {
+		documentManagers.get(editor.document.uri)?.parseCells();
+	}
+}
+
+export function getOrCreateDocumentManager(editor: vscode.TextEditor) {
+	let docManager = documentManagers.get(editor.document.uri);
+	if (!docManager) {
+		docManager = new DocumentManager(editor);
+		docManager.parseCells();
+	}
+	return docManager;
+}
+
+// Creates documentManager and parses cells, if needed
+export function setupDocumentManager(editor: vscode.TextEditor | undefined) {
+	if (editor && canHaveCells(editor.document)) {
+		getOrCreateDocumentManager(editor);
+	}
+}
+
+export function getActiveDocumentManager(): DocumentManager | undefined {
+	const activeEditor = vscode.window?.activeTextEditor;
+	if (activeEditor && canHaveCells(activeEditor.document)) {
+		return getOrCreateDocumentManager(activeEditor);
+	}
+
+	return undefined;
+}
+
+export function getEditorFromDocument(document: vscode.TextDocument): vscode.TextEditor | undefined {
+	return vscode.window.visibleTextEditors.find(
+		(editor) => editor.document === document
+	);
+}
```

Comment:
Referring to my first comment, this function should not exist, since documents and editors have a one-to-many relationship.

We'll be able to remove it if we make `DocumentManager.constructor` take a `document` instead of an `editor`, and if we add a new class that handles all of the cell commands. This new class would have the following public methods:

- `constructor`: accepting an `editor` – then getting `editor.document`'s `DocumentManager` and getting/parsing its cells as needed
- `runAllCells`
- `runCurrentCell`
- `runCurrentAdvance`
- `runPreviousCell`
- `runNextCell`
- `runCellsAbove`
- `runCurrentAndBelow`
- `runCellsBelow`
- `goToPreviousCell`
- `goToNextCell`
- `insertCodeCell`